File: src/test/scala/com/exasol/cloudetl/kafka/ittests/RecordFieldSpecificationIT.scala
package com.exasol.cloudetl.kafka

import java.lang.{Integer => JInt}
import java.lang.{Long => JLong}

import com.exasol.ExaMetadata
import com.exasol.common.json.JsonMapper

import com.fasterxml.jackson.databind.JsonNode
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.serialization.StringSerializer
import org.mockito.ArgumentCaptor
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.Mockito.when
import org.mockito.invocation.InvocationOnMock
import org.mockito.stubbing.Answer

/**
 * A test class that tests the {@code RECORD_FIELDS} integration and not the offset and partition handling.
 */
class RecordFieldSpecificationIT extends KafkaTopicDataImporterAvroIT {

  private[this] val customRecord = AvroRecord("abc", 3, 13)
  implicit val stringSerializer = new StringSerializer

  test("default must be 'value.*': All fields from the record") {
    createCustomTopic(topic)
    publishToKafka(topic, customRecord)
    assert(getEmittedValues("value.*", Seq(classOf[String], classOf[JInt], classOf[JLong])) === Seq("abc", 3, 13))
  }

  test("must emit multiple record value fields in the order specified") {
    createCustomTopic(topic)
    publishToKafka(topic, customRecord)
    assert(getEmittedValues("value.col_long, value.col_str", Seq(classOf[JLong], classOf[String])) === Seq(13, "abc"))
  }

  test("must be able to reference the full value") {
    createCustomTopic(topic)
    publishToKafka(topic, customRecord)
    val result = getEmittedValues("value", Seq(classOf[String]))
    assert(result.size === 1)
    assertJson(
      result(0).asInstanceOf[String],
      """ {
        |   "col_str": "abc",
        |   "col_int": 3,
        |   "col_long": 13
        | }
        |""".stripMargin
    )
  }

  test("must be able to reference key values with default RECORD_KEY_FORMAT string") {
    createCustomTopic(topic)
    publishToKafka(topic, "string_key", customRecord)
    assert(getEmittedValues("key, value.col_long", Seq(classOf[String], classOf[JLong])) === Seq("string_key", 13))
  }

  test("must fail when the key is accessed with concrete field") {
    createCustomTopic(topic)
    publishToKafka(topic, "string_key", customRecord)
    intercept[KafkaConnectorException] {
      getEmittedValues("key.someFieldReference", Seq(classOf[String]))
    }
  }

  test("must fail when the key is accessed with the 'all fields' reference") {
    createCustomTopic(topic)
    publishToKafka(topic, "string_key", customRecord)
    intercept[KafkaConnectorException] {
      getEmittedValues("key.*", Seq())
    }
  }

  test("must handle null key and values") {
    createCustomTopic(topic)
    publishToKafka(topic, null.asInstanceOf[String], null.asInstanceOf[AvroRecord])
    assert(getEmittedValues("key, value", Seq(classOf[String], classOf[Any])) === Seq(null, null))
  }

  test("must handle null values combined with a present key and timestamps") {
    createCustomTopic(topic)
    publishToKafka(topic, "theKey", null.asInstanceOf[AvroRecord])
    val values = getEmittedValues("key, timestamp, value", Seq(classOf[String], classOf[JLong], classOf[Any]))
    assert(values.size === 3)
    assert(values(0) === "theKey")
    assert(values(1).isInstanceOf[JLong])
    assert(values(2) === null)
  }

  test("must include record timestamp when in fieldlist") {
    createCustomTopic(topic)
    val recordTimestamp = 123456L
    withProducer[String, AvroRecord, Unit] { producer =>
      producer.send(new ProducerRecord[String, AvroRecord](topic, 0, recordTimestamp, "record_key", customRecord))
      ()
    }
    val values = getEmittedValues("timestamp, value.str_col, key", Seq(classOf[JLong], classOf[Any], classOf[String]))
    assert(values === Seq(recordTimestamp, null, "record_key"))
  }

  private[this] def assertJson(actual: String, expected: String): Unit = {
    assert(JsonMapper.fromJson[JsonNode](actual) === JsonMapper.fromJson[JsonNode](expected))
    ()
  }

  private[this] def getEmittedValues(recordFieldsStmt: String, outputColumnTypes: Seq[Class[_]]): Seq[Any] = {
    val mockedIterator = mockExasolIterator(properties ++ Map("RECORD_FIELDS" -> recordFieldsStmt), Seq(0), Seq(-1))
    val outputColumnTypesWithMeta = outputColumnTypes ++ Seq(classOf[JInt], classOf[JLong])
    val columnCount = outputColumnTypesWithMeta.size
    val exasolMetadata = mock[ExaMetadata]
    when(exasolMetadata.getOutputColumnCount()).thenReturn(columnCount)
    when(exasolMetadata.getOutputColumnType(anyInt())).thenAnswer(new Answer[Class[_]]() {
      override def answer(invocation: InvocationOnMock): Class[_] = {
        val columnIndex = invocation.getArguments()(0).asInstanceOf[JInt]
        outputColumnTypesWithMeta(columnIndex)
      }
    })
    KafkaTopicDataImporter.run(exasolMetadata, mockedIterator)

    val captor = ArgumentCaptor.forClass(classOf[Array[Any]])
    verify(mockedIterator, times(1)).emit(captor.capture())

    val valuesEmitted = captor.getValue().toSeq
    // Don't need the last two values (partition and offset)
    valuesEmitted.slice(0, valuesEmitted.size - 2)
  }

}

File: src/test/java/com/exasol/cloudetl/kafka/RecordFieldSpecificationIT.java
package com.exasol.cloudetl.kafka;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.*;

import java.util.*;

import org.apache.kafka.clients.producer.ProducerRecord;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;

import com.exasol.ExaMetadata;

class RecordFieldSpecificationIT extends KafkaTopicDataImporterAvroIT {
    private final AvroRecord customRecord = new AvroRecord("abc", 3, 13);

    @Test
    void defaultIsAllFieldsFromRecordValue() throws Exception {
        createCustomTopic(this.topic);
        publishAvro(this.topic, this.customRecord);
        assertEquals(List.of("abc", 3, 13L), getEmittedValues("value.*", List.of(String.class, Integer.class, Long.class)));
    }

    @Test
    void emitsMultipleRecordValueFieldsInSpecifiedOrder() throws Exception {
        createCustomTopic(this.topic);
        publishAvro(this.topic, this.customRecord);
        assertEquals(List.of(13L, "abc"), getEmittedValues("value.col_long, value.col_str",
                List.of(Long.class, String.class)));
    }

    @Test
    void referencesFullValue() throws Exception {
        createCustomTopic(this.topic);
        publishAvro(this.topic, this.customRecord);
        final List<Object> result = getEmittedValues("value", List.of(String.class));
        assertEquals(1, result.size());
        assertJson((String) result.get(0), "{\"col_str\":\"abc\",\"col_int\":3,\"col_long\":13}");
    }

    @Test
    void referencesKeyValuesWithDefaultRecordKeyFormatString() throws Exception {
        createCustomTopic(this.topic);
        publishAvro(this.topic, "string_key", this.customRecord);
        assertEquals(List.of("string_key", 13L),
                getEmittedValues("key, value.col_long", List.of(String.class, Long.class)));
    }

    @Test
    void failsWhenKeyIsAccessedWithConcreteField() {
        createCustomTopic(this.topic);
        publishAvro(this.topic, "string_key", this.customRecord);
        final List<Class<?>> outputColumnTypes = List.of(String.class);
        assertThrows(KafkaConnectorException.class, () -> getEmittedValues("key.someFieldReference", outputColumnTypes));
    }

    @Test
    void failsWhenKeyIsAccessedWithAllFieldsReference() {
        createCustomTopic(this.topic);
        publishAvro(this.topic, "string_key", this.customRecord);
        assertThrows(KafkaConnectorException.class, () -> getEmittedValues("key.*", List.of()));
    }

    @Test
    void handlesNullKeyAndValues() throws Exception {
        createCustomTopic(this.topic);
        publishAvro(this.topic, null, null);
        assertEquals(Arrays.asList(null, null), getEmittedValues("key, value", List.of(String.class, Object.class)));
    }

    @Test
    void handlesNullValuesCombinedWithPresentKeyAndTimestamps() throws Exception {
        createCustomTopic(this.topic);
        publishAvro(this.topic, "theKey", null);
        final List<Object> values = getEmittedValues("key, timestamp, value",
                List.of(String.class, Long.class, Object.class));
        assertAll(() -> assertEquals(3, values.size()),
                () -> assertEquals("theKey", values.get(0)),
                () -> assertInstanceOf(Long.class, values.get(1)),
                () -> assertNull(values.get(2)));
    }

    @Test
    void includesRecordTimestampWhenInFieldList() throws Exception {
        createCustomTopic(this.topic);
        final long recordTimestamp = 123456L;
        publishToKafka(new ProducerRecord<>(this.topic, 0, recordTimestamp, "record_key", this.customRecord),
                avroSerializer());
        assertEquals(Arrays.asList(recordTimestamp, null, "record_key"),
                getEmittedValues("timestamp, value.str_col, key", List.of(Long.class, Object.class, String.class)));
    }

    private void assertJson(final String actual, final String expected) {
        assertEquals(JsonArgumentMatcher.readJson(expected), JsonArgumentMatcher.readJson(actual));
    }

    private List<Object> getEmittedValues(final String recordFieldsStmt, final List<Class<?>> outputColumnTypes)
            throws Exception {
        final Map<String, String> newProperties = new LinkedHashMap<>(this.properties);
        newProperties.put("RECORD_FIELDS", recordFieldsStmt);
        final var iterator = mockExasolIterator(newProperties, List.of(0), List.of(-1L));
        final List<Class<?>> outputColumnTypesWithMeta = new ArrayList<>(outputColumnTypes);
        outputColumnTypesWithMeta.add(Integer.class);
        outputColumnTypesWithMeta.add(Long.class);
        final ExaMetadata metadata = mock(ExaMetadata.class);
        when(metadata.getOutputColumnCount()).thenReturn((long) outputColumnTypesWithMeta.size());
        when(metadata.getOutputColumnType(anyInt())).thenAnswer(invocation -> outputColumnTypesWithMeta.get(invocation.getArgument(0)));
        KafkaTopicDataImporter.run(metadata, iterator);

        final ArgumentCaptor<Object[]> captor = ArgumentCaptor.forClass(Object[].class);
        verify(iterator).emit(captor.capture());
        final Object[] values = captor.getValue();
        return Arrays.asList(Arrays.copyOf(values, values.length - 2));
    }
}
